File: leetcode/2001-2999/2352-equal-row-and-column-pairs/EqualRowAndColumnPairs.ts
function equalPairs(grid: number[][]): number {
  const n = grid.length;

  const equals = (row: number, column: number) => {
    for (let i = 0; i < n; i++) {
      const cr: number = grid[row][i];
      const cl: number = grid[i][column];

      if (cr !== cl) {
        return false;
      }
    }

    return true;
  }

  let totalEqualPairs = 0;

  for (let i = 0; i < n; i++) {
    for (let j = 0; j < n; j++) {
      totalEqualPairs += equals(i, j) ? 1 : 0;
    }
  }

  return totalEqualPairs;
};

console.log(equalPairs([[3, 1, 2, 2], [1, 4, 4, 5], [2, 4, 2, 2], [2, 4, 2, 2]]))

File: leetcode/2001-3000/2301-2400/2351-2360/2352-equal-row-and-column-pairs/EqualRowAndColumnPairs.java
import java.util.Arrays;

public class EqualRowAndColumnPairs {

  public static void main(String[] args) {
    // int[][] grid = {
    // { 3, 2, 1 },
    // { 1, 7, 6 },
    // { 2, 7, 7 }
    // };

    int[][] grid = { { 3, 1, 2, 2 }, { 1, 4, 4, 5 }, { 2, 4, 2, 2 }, { 2, 4, 2, 2 } };

    System.out.println(new Solution().equalPairs(grid));
  }
}

class Solution {
  public int equalPairs(int[][] grid) {
    int n = grid.length;
    int[][] transposed = new int[n][n];

    int totalEqualPairs = 0;

    for (int i = 0; i < n; i++) {
      for (int j = 0; j < n; j++) {
        transposed[i][j] = grid[j][i];
      }
    }

    for (int[] row : grid) {
      for (int[] column : transposed) {
        totalEqualPairs += Arrays.equals(row, column) ? 1 : 0;
      }
    }

    return totalEqualPairs;
  }
}
